cpp/include/cudf/join/sort_merge_join.hpp

```diff
@@ -0,0 +1,254 @@
+   * the result of performing an inner join between two tables
+   */
+  std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+            std::unique_ptr<rmm::device_uvector<size_type>>>
+  inner_join(rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+             rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref());
+
+ private:
+  /**
+   * @brief Helper struct to pre-process tables before join operations
+   */
+  struct preprocessed_table {
+    table_view _tbl_view;  ///< Unprocessed table view before pre-processing
+
+    table_view
+      _null_processed_tbl_view;  ///< Processed table view which is the null-free subset of the rows
+                                 ///< of the unprocessed table view if null equality is set to
+                                 ///< false, otherwise equal to the unprocessed table view
+
+    std::optional<rmm::device_buffer> _validity_mask =
+      std::nullopt;  ///< Optional validity mask for null_equality::UNEQUAL case
+    std::optional<size_type> _num_nulls =
+      std::nullopt;  ///< Optional count of nulls for null_equality::UNEQUAL case
+    std::optional<std::unique_ptr<table>> _null_processed_tbl =
+      std::nullopt;  ///< Optional filtered table for null_equality::UNEQUAL case
+
+    std::optional<std::unique_ptr<column>> _null_processed_tbl_sorted_order =
+      std::nullopt;  ///< Optional sort ordering for pre-sorted tables
+
+    /**
+     * @brief Mark rows in unprocessed table with nulls at root or child levels by populating the
+     * _validity_mask
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     */
+    void populate_nonnull_filter(rmm::cuda_stream_view stream);
+
+    /**
+     * @brief Apply _validity_mask to the _tbl_view to create a null-free table
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     */
+    void apply_nonnull_filter(rmm::cuda_stream_view stream);
+
+    /**
+     * @brief Pre-process the unprocessed table when null equality is set to unequal
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     */
+    void preprocess_unprocessed_table(rmm::cuda_stream_view stream);
+
+    /**
+     * @brief Compute sorted ordering of the processed table
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     */
+    void get_sorted_order(rmm::cuda_stream_view stream);
+
+    /**
+     * @brief Create mapping from processed table indices to unprocessed table indices
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     * @return A device vector containing the mapping from processed table indices to unprocessed
+     * table indices
+     */
+    rmm::device_uvector<size_type> map_tbl_to_unprocessed(rmm::cuda_stream_view stream);
```

Comment:
Let's avoid abbreviations.
```suggestion
    rmm::device_uvector<size_type> map_table_to_unprocessed(rmm::cuda_stream_view stream);
```